### PR TITLE
[backend] use common error types

### PR DESCRIPTION
**Pull request by freben**
Adds common error types that backends can throw, and that the error middleware will know how to handle sensible and translate to the proper HTTP response codes.

### CODE REVIEW

File: packages/backend-common/src/errors.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * The request is malformed and cannot be processed.
+ */
+export class BadRequestError extends Error {
+  readonly cause?: Error;
+
+  constructor(message?: string, cause?: Error) {
+    super(message);
+    Object.setPrototypeOf(this, BadRequestError.prototype);
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:36]: Better do all of this stuff in a common CustomError base class instead.

+1 for keeping track of cause, I think we could even do something like ``super(`${message}, ${cause}`)`` when there is a cause.

Comment (freben) [reply packages/backend-common/src/errors.ts:36]: But does one not have to actually give the subclass type here? I am not sure if it CAN be delegated to a base class.

Comment (freben) [reply packages/backend-common/src/errors.ts:36]: Oh like `class X extends ErrorBase { constructor(...) { super(X); } }`?

Comment (freben) [reply packages/backend-common/src/errors.ts]: Fixed

File: packages/backend-common/src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  constructor(constructor: Function, message?: string, cause?: Error) {
+    super(message);
+    Object.setPrototypeOf(this, constructor.prototype);
+    Error.captureStackTrace(this, constructor);
+    this.name = this.constructor.name;
+    this.cause = cause;
+  }
+
+  toString() {
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:39]: I think passing the message created here on to super would be better. With this you'll get different behaviour depending on how the error is used. For example the stack trace won't include the cause, or someone using `error.message` to print the error without the constructor name prefix.

File: packages/backend-common/src/errors.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * use e.g. the http-errors library.
+ */
+
+class CustomErrorBase extends Error {
+  readonly cause?: Error;
+
+  constructor(constructor: Function, message?: string, cause?: Error) {
+    super(message);
+    Object.setPrototypeOf(this, constructor.prototype);
+    Error.captureStackTrace(this, constructor);
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:34]: what's the difference between `constructor` and `this.constructor`? :grin:

File: packages/backend-common/src/errors.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ * intended to be thrown by the request handling layer. In those places, please
+ * use e.g. the http-errors library.
+ */
+
+class CustomErrorBase extends Error {
+  readonly cause?: Error;
+
+  constructor(constructor: Function, message?: string, cause?: Error) {
+    super(message);
+    Object.setPrototypeOf(this, constructor.prototype);
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:33]: Why is this needed?

File: packages/backend-common/src/errors.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+export class UnauthenticatedError extends CustomErrorBase {
+  constructor(message?: string, cause?: Error) {
+    super(UnauthenticatedError, message, cause);
+  }
+}
+
+/**
+ * The authenticated caller is not permitted to perform this request.
+ */
+export class ForbiddenError extends CustomErrorBase {
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:71]: I prefer to call this `NotAllowedError`

File: packages/backend-common/src/errors.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+export class BadRequestError extends CustomErrorBase {
+  constructor(message?: string, cause?: Error) {
+    super(BadRequestError, message, cause);
+  }
+}
+
+/**
+ * The request requires authentication, which was not properly supplied.
+ */
+export class UnauthenticatedError extends CustomErrorBase {
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:62]: I prefer to call this `AuthenticationError`

File: packages/backend-common/src/errors.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    return result;
+  }
+}
+
+/**
+ * The request is malformed and cannot be processed.
+ */
+export class BadRequestError extends CustomErrorBase {
```
Comment (Rugvip) [inline packages/backend-common/src/errors.ts:53]: Not sure if we should have this tbh, might be leaking a bit too much http, but fine with it for now at least.

File: packages/backend-common/src/errors.ts (hunk 3 above)
Comment (freben) [reply packages/backend-common/src/errors.ts:34]: I think you'll actually have to tell me 🤔

File: packages/backend-common/src/errors.ts (hunk 4 above)
Comment (freben) [reply packages/backend-common/src/errors.ts:33]: It's due to a breaking change that was introduced some time ago in typescript. It actually broke without it.

https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

File: packages/backend-common/src/errors.ts (hunk 6 above)
Comment (freben) [reply packages/backend-common/src/errors.ts:62]: Sounds like an authentication attempt failed :) But ok

File: packages/backend-common/src/errors.ts (hunk 5 above)
Comment (freben) [reply packages/backend-common/src/errors.ts:71]: Kk

File: packages/backend-common/src/errors.ts (hunk 2 above)
Comment (freben) [reply packages/backend-common/src/errors.ts:39]: Is it ok to "mess with" the error's message like that? Sure I could change that.

Oooor should I remove the cause functionality?

File: packages/backend-common/src/errors.ts (hunk 3 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts:34]: Does it work if we bump the TS target instead?

File: packages/backend-common/src/errors.ts (hunk 4 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts:33]: Does it work if we bump the TS target instead?

File: packages/backend-common/src/errors.ts (hunk 3 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts:34]: I don't think there should be one, but maybe there actually is due to the TS builtin class issue. Still interested in if bumping the TS target works

File: packages/backend-common/src/errors.ts (hunk 2 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts:39]: I mean you're already messing with it? :p

Yeah we could skip it for now, but I usually end up doing it manually in the message, and losing information in the process since only the message of the cause is kept.

File: packages/backend-common/src/errors.ts (hunk 6 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts:62]: Yeah and I'd use it for both, as in both login failed because of bad password and request failed because access token is invalid.

File: packages/backend-common/src/errors.ts (hunk 4 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: It did!

File: packages/backend-common/src/errors.ts (hunk 6 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: Okeydokey!

File: packages/backend-common/src/errors.ts (hunk 7 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: Yeah i just want SOME form of path for reporting that the input was invalid (something that ultimately becomes a 400).

File: packages/backend-common/src/errors.ts (hunk 3 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: Yassss it could be removed now and with the modern target!

File: packages/backend-common/src/errors.ts (hunk 4 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts]: 🎉

File: packages/backend-common/src/errors.ts (hunk 2 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: Chose to retain it, yeah it can be useful

File: packages/backend-common/src/errors.ts (hunk 7 above)
Comment (Rugvip) [reply packages/backend-common/src/errors.ts]: Yup, thinking about alternatives like `InputError` or `BadValueError`

**Review by Rugvip — APPROVED** [review]

File: packages/backend-common/src/errors.ts (hunk 7 above)
Comment (freben) [reply packages/backend-common/src/errors.ts]: Wanna go for `InputError`? I'm game!

Comment (Rugvip) [reply packages/backend-common/src/errors.ts]: image

Comment (freben) [reply packages/backend-common/src/errors.ts]: done
